### ISSUE TITLE
[python-package] early_stopping with min_delta has a minor bug I think

### ISSUE THREAD

**Issue by xenoverse-us** [microsoft/LightGBM#6123]
## Description
When setting early_stopping with min_delta, it seems lightgbm drops iterations between the last best and the iterations that ran so far without surpassing the min_delta.

Example here:
```
Training until validation scores don't improve for 1000.0 rounds
[1000]  cv_agg's valid auc: 0.903716 + 0.000331755
[2000]  cv_agg's valid auc: 0.912798 + 0.000420065
[3000]  cv_agg's valid auc: 0.917164 + 0.000331079
[4000]  cv_agg's valid auc: 0.920226 + 0.000328495
[5000]  cv_agg's valid auc: 0.922254 + 0.000342766
[6000]  cv_agg's valid auc: 0.924017 + 0.000306959
[7000]  cv_agg's valid auc: 0.925224 + 0.000166862
[8000]  cv_agg's valid auc: 0.926138 + 0.000138658
Early stopping, best iteration is:
[7109]  cv_agg's valid auc: 0.925321 + 0.000168736
```
As you can see, it claims iteration 7109 is the best, with auc 0.925321, runs after that until 8xxx (I only print log every boost_round /10) has produced better metric such as 0.926138, but it is below the `min_delta` threshold of 0.926321 (previous best `0.925321` + `0.001`). Yet this is still better than the previous best so I would've expected this to be returned and used instead of iteration 7109.

## Reproducible example
```
boost_round = 10000
tuner = optuna.integration.LightGBMTunerCV(param,
                            dtrain,
                            study=study,
                            verbose_eval=False,
                            seed = 42,
                            folds=stratified_k_fold,
                            num_boost_round=boost_round,
                            callbacks=[
                                early_stopping(boost_round / 10, min_delta=0.001),
                                log_evaluation(boost_round / 10),
                            ],
                            model_dir=study_name,
                            return_cvbooster=True,
                            optuna_callbacks=[trial_callback],
                            time_budget=3600 * 24 * 7
                           )
```

## Environment info
lightgbm 4.1.0

```shell
pip install lightgbm
```

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

> Yet this is still better than the previous best so I would've expected this to be returned and used instead of iteration 7109.

This is intended behavior, not a bug.

`min_delta` tells LightGBM to treat an improvement of `< min_delta` in an eval metric as if it was not an improvement at all, and therefore to consider such iterations not "better" than previous ones. That's an overfitting protection.

If you think that a new tree that led to an improvement of `< min_delta` in some metric makes a model more desirable than one without that tree, you should choose a smaller value of `min_delta`.

----

> minor bug

We really appreciate you taking the time to report behavior that you believe to be a bug!

But in the future, please try to create a minimal, reproducible example that does not depend on higher-level interfaces to LightGBM like `optuna` and which we can run and observe the same behavior ([docs on what that means](https://stackoverflow.com/help/minimal-reproducible-example)).

For example, you haven't told us some important details that'd be necessary to reproduce what you're seeing:

* what data you're using
* the content of `params` (what metrics are you using, for example?)
* what `trial_callback` is and does
* what `boost_round` is (and therefore, how many early stopping rounds you're using)
